src/python/pants/core/goals/build.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import logging
+from abc import ABCMeta
+from dataclasses import dataclass
+from typing import Optional
+
+from pants.core.util_rules.distdir import DistDir
+from pants.engine.fs import Digest, MergeDigests, Snapshot, Workspace
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.rules import Get, MultiGet, collect_rules, goal_rule
+from pants.engine.target import FieldSet, TargetRootsToFieldSets, TargetRootsToFieldSetsRequest
+from pants.engine.unions import union
+
+logger = logging.getLogger(__name__)
+
+
+@union
+class BuildFieldSet(FieldSet, metaclass=ABCMeta):
+    """The fields necessary to build an asset from a target."""
+
+
+@dataclass(frozen=True)
+class BuiltAsset:
+    digest: Digest
+    relpath: str
+    extra_log_info: Optional[str] = None
+
+
+class BuildSubsystem(GoalSubsystem):
+    """Build an asset, such as an archive, JAR, PEX, AWS Lambda, etc."""
```

Comment:
Ah, this is the "asset" definition I was looking for. I think my earlier comments can be ignored since this will show up in `./pants help`.

Reply:
Let me put up a change to rename this all to `package` :) I really like JP and your feedback on that rename, rather than `build`.